File: project_euler/02/024.py
#!/usr/bin/env python3
'''Copyright [2017] <dmnsn7@gmail.com>'''

N = 10
K = 1000000


def kth_permutation(num_n, num_k):
    fac = [1 for _ in range(num_n)]
    for i in range(1, num_n):
        fac[i] = fac[i - 1] * i

    permutation = []
    is_used = [0 for _ in range(num_n)]
    for i in range(num_n):
        kth = num_k // fac[num_n - i - 1]
        num_k %= fac[num_n - i - 1]
        cnt = 0
        for j in range(num_n):
            cnt += 1 if not is_used[j] else 0
            if cnt == kth + 1:
                permutation.append(j)
                is_used[j] = 1
                break

    return permutation


def main():
    permutation = kth_permutation(N, K - 1)
    for _, dig in enumerate(permutation):
        print(dig, end='')

    print('')


if __name__ == '__main__':
    main()

File: project_euler/02/024.go
// Copyright [2017] <dmnsn7@gmail.com>

package main

import "fmt"

const cn = 10
const ck = 1000000

func kthPermutation(n, k int) []int {
	fac := make([]int, n)
	for i := 0; i < n; i++ {
		if i == 0 {
			fac[i] = 1
		} else {
			fac[i] = fac[i-1] * i
		}
	}

	permutation := []int{}
	isUsed := make([]bool, n)
	for i := 0; i < n; i++ {
		isUsed[i] = false
	}

	for i := 0; i < n; i++ {
		kth := k / fac[n-i-1]
		k %= fac[n-i-1]
		for j, cnt := 0, 0; j < n; j++ {
			if !isUsed[j] {
				cnt++
			}
			if cnt == kth+1 {
				permutation = append(permutation, j)
				isUsed[j] = true
				break
			}
		}
	}

	return permutation
}

func main() {
	permutation := kthPermutation(cn, ck-1)
	for i := 0; i < len(permutation); i++ {
		fmt.Printf("%c", permutation[i]+'0')
	}

	fmt.Printf("\n")
}
